HDDS-11367. Improve ozone balancing status command output

juncevich opened the pull request:
## What changes were proposed in this pull request?
Minor improvements to command output:
1. job task time consumption
2. iteration time consumption
3. unfinished iteration shouldn't be in iteration history section
4. show data units, even it less than gb show it in mb

## What is the link to the Apache JIRA
[HDDS-11367](https://issues.apache.org/jira/browse/HDDS-11367)

## How was this patch tested?
Tested by robot test

siddhantsangwan reviewed (commented):
@juncevich can you please elaborate on what these points mean?

> job task time consumption
iteration time consumption

It will help reviewers understand what you're proposing.

juncevich commented:
> @juncevich can you please elaborate on what these points mean?
> 
> > job task time consumption
> > iteration time consumption
> 
> It will help reviewers understand what you're proposing.

Hi, @siddhantsangwan. Sure, i can explain. Balancing job constists from some iterations. Job task time - the whole balancing time, iteration time - particular iteration time. Job time  = iteration time + iteration time ...

afilpp commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ContainerBalancerStatusSubcommand.java:
```diff
@@ -159,20 +166,46 @@ private String getPrettyIterationStatusInfo(ContainerBalancerTaskIterationStatus
+
+  public static String getPrettySize(long sizeInBytes) {
+    if (sizeInBytes / OzoneConsts.GB > 0) {
+      return sizeInBytes / OzoneConsts.GB + " Gb" + sizeInBytes % OzoneConsts.GB / OzoneConsts.MB + " Mb";
```
Maybe add a space here to avoid displaying something like - "4 GB100 MB"?

juncevich commented:
Hi [siddhantsangwan](https://github.com/siddhantsangwan). Can you review my PR?

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -340,12 +344,13 @@ public List<ContainerBalancerTaskIterationStatusInfo> getCurrentIterationsStatis
         .mapToInt(ContainerBalancerTaskIterationStatusInfo::getIterationNumber)
         .max()
         .orElse(0);
```
Wouldn't the last index in the list be the latest iteration? Why do we need to scan through it?

siddhantsangwan reviewed (changes requested):
@juncevich thanks for working on this. Other than my comments below, we definitely need more testing here. I'd like to at least see:
1. Unit tests for balancer that verify the iteration duration and other properties are correct. 
2. Tests for the balancer status command that verify the code is correctly picking up the mocked values (mocked response from SCM).

We can also add verifications to the robot tests.

juncevich commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ContainerBalancerStatusSubcommand.java:
```diff
@@ -159,20 +166,46 @@ private String getPrettyIterationStatusInfo(ContainerBalancerTaskIterationStatus
             "Exited data from nodes", leavingDataNodeList);
   }
+
+  private String getPrettyIterationStatusInfo(long duration) {
```
I didn't find utility methods. I will extract this method to separate util class and write test for it.

juncevich commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ContainerBalancerStatusSubcommand.java:
```diff
@@ -78,7 +85,9 @@ public void execute(ScmClient scmClient) throws IOException {
           System.out.println("Iteration history list:");
           System.out.println(
-              iterationsStatusInfoList.stream().map(this::getPrettyIterationStatusInfo)
+              iterationsStatusInfoList.subList(0, iterationsStatusInfoList.size() - 1)
```
I need iterations without current.

>Sublist call feels like absolutely redundant.
Why you think so?

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -307,53 +311,53 @@ private void balance() {
-        .orElse(0);
-
+    int lastIterationNumber = iterationsStatistic.isEmpty() ? 0
+        : iterationsStatistic.get(iterationsStatistic.size() - 1).getIterationNumber();
```
It doesn't work. This is not variable.

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -118,7 +121,7 @@ public class ContainerBalancerTask implements Runnable {
   private boolean delayStart;
   private List<ContainerBalancerTaskIterationStatusInfo> iterationsStatistic;
-
+  private OffsetDateTime currentIterationStarted;
```
Sorry, but I can't see any differences here.

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -307,53 +311,53 @@ private void balance() {
+        iR.name(),
+        iterationDuration,
+        getSizeScheduledForMoveInLatestIteration(),
+        metrics.getDataSizeMovedInLatestIteration(),
```
Can you describe this notice? This is not obvious now.

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -118,7 +121,7 @@ public class ContainerBalancerTask implements Runnable {
   private boolean delayStart;
   private List<ContainerBalancerTaskIterationStatusInfo> iterationsStatistic;
-
+  private OffsetDateTime currentIterationStarted;
```
It looks like other fields in this class. I think only one thread will get access to this field because it is located in Runnable.
<img width="720" alt="image" src="https://github.com/user-attachments/assets/ed203b1f-28dc-4b5f-a746-8a0b2a2b4bdb">

adoroszlai commented:
@juncevich can you please check test timeouts / failures?

```
org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit
org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerStatusInfo
org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerTask
```

https://github.com/juncevich/ozone/actions/runs/11380961580/artifacts/2067990610

juncevich commented:
> @juncevich can you please check test timeouts / failures?
> 
> ```
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerDatanodeNodeLimit
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerStatusInfo
> org.apache.hadoop.hdds.scm.container.balancer.TestContainerBalancerTask
> ```
> 
> https://github.com/juncevich/ozone/actions/runs/11380961580/artifacts/2067990610

Thanks for the comment. Tests fixed.

adoroszlai commented:
@siddhantsangwan can you please take another look?

siddhantsangwan commented:
@juncevich Is this ready for another round of reviews?

juncevich commented:
> @juncevich Is this ready for another round of reviews?

Yes, I try to fix the review notices ASAP because they are blocking another fix. I didn’t bring it up because it’s a different task

juncevich commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ContainerBalancerStatusSubcommand.java:
```diff
@@ -159,20 +169,23 @@ private String getPrettyIterationStatusInfo(ContainerBalancerTaskIterationStatus
                     "%-50s %n%s",
             "Key", "Value",
-            "Iteration number", iterationNumber,
+            "Iteration number", iterationNumber == 0 ? "-" : iterationNumber,
```
I don't understand why it's bad and how I should refactor it.

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTaskIterationStatusInfo.java:
```diff
@@ -18,86 +18,184 @@
+  private final Map<UUID, Long> sizeLeavingNodes;
 
   @SuppressWarnings("checkstyle:ParameterNumber")
   public ContainerBalancerTaskIterationStatusInfo(
```
I think in some cases it is OK.
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/807c907c-ab5e-45a0-a7a0-4f6f279a81ad">

juncevich commented on hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/datanode/TestContainerBalancerSubCommand.java:
```diff
@@ -206,27 +178,298 @@ public void testContainerBalancerStatusInfoSubcommandRunning()
+        "Exited data from nodes                             \n" +
+        "b8b9c511-c30f-4933-8938-2f272e307070 -> 30 GB\n" +
+        "7bd99815-47e7-4015-bc61-ca6ef6dfd130 -> 18 GB";
+    assertFalse(output.contains(currentIterationOutput));
```
Why? It is common practice to use assertFalse. If this is a problem, let’s refactor it at the project level in a separate task.
<img width="1500" alt="image" src="https://github.com/user-attachments/assets/5f0996a1-71fd-49eb-81e8-d476abb8a549">

juncevich commented:
@ivanzlenko No need to use Atomic for OffsetDateTime, because it thread-safe by design.
<img width="699" alt="image" src="https://github.com/user-attachments/assets/8cd2ec50-05bd-47d8-9ac4-2df5b9f8ad1c">

juncevich commented:
@siddhantsangwan, can you take another look at this PR?

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/IterationInfo.java:
```diff
@@ -0,0 +1,47 @@
+
+package org.apache.hadoop.hdds.scm.container.balancer;
+
+/**
+ * Information about the process of moving data.
+ */
```
Looks like you forgot to change this java doc, it's currently not relevant.

siddhantsangwan commented on hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/datanode/TestContainerBalancerSubCommand.java:
```diff
@@ -206,27 +178,298 @@ public void testContainerBalancerStatusInfoSubcommandRunning()
+
+    String currentIterationOutput =
+        "Current iteration info:\n" +
+        "Key                                                Value\n" +
+        "Iteration number                                   3\n" +
+        "Iteration duration                                 1h 6m 40s\n" +
+        "Iteration result                                   IN_PROGRESS\n" +
+        "Size scheduled to move                             48 GB\n" +
+        "Moved data size                                    48 GB\n" +
+        "Scheduled to move containers                       11\n" +
+        "Already moved containers                           11\n" +
+        "Failed to move containers                          0\n" +
+        "Failed to move containers by timeout               0\n" +
+        "Entered data to nodes                              \n" +
+        "80f6bc27-e6f3-493e-b1f4-25f810ad960d <- 20 GB\n" +
+        "701ca98e-aa1a-4b36-b817-e28ed634bba6 <- 28 GB\n" +
+        "Exited data from nodes                             \n" +
+        "b8b9c511-c30f-4933-8938-2f272e307070 -> 30 GB\n" +
+        "7bd99815-47e7-4015-bc61-ca6ef6dfd130 -> 18 GB";
+    assertTrue(output.contains(currentIterationOutput));
+
+    assertTrue(output.contains("Iteration history list:"));
+    String firstHistoryIterationOutput =
+        "Key                                                Value\n" +
+        "Iteration number                                   3\n" +
+        "Iteration duration                                 1h 6m 40s\n" +
+        "Iteration result                                   IN_PROGRESS\n" +
+        "Size scheduled to move                             48 GB\n" +
+        "Moved data size                                    48 GB\n" +
+        "Scheduled to move containers                       11\n" +
+        "Already moved containers                           11\n" +
+        "Failed to move containers                          0\n" +
+        "Failed to move containers by timeout               0\n" +
+        "Entered data to nodes                              \n" +
+        "80f6bc27-e6f3-493e-b1f4-25f810ad960d <- 20 GB\n" +
+        "701ca98e-aa1a-4b36-b817-e28ed634bba6 <- 28 GB\n" +
+        "Exited data from nodes                             \n" +
+        "b8b9c511-c30f-4933-8938-2f272e307070 -> 30 GB\n" +
+        "7bd99815-47e7-4015-bc61-ca6ef6dfd130 -> 18 GB";
+    assertTrue(output.contains(firstHistoryIterationOutput));
```
Are we asserting the same thing twice? It looks like `currentIterationOutput` and `firstHistoryIterationOutput` have the same content.

juncevich replied:
Thanks! I will fix the test data. The output between the current and historical iterations should be different.

siddhantsangwan commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/ContainerBalancerStatusSubcommand.java:
```diff
@@ -60,12 +64,15 @@ public void execute(ScmClient scmClient) throws IOException {
-        System.out.printf("Started at: %s %s%n%n", dateTime.toLocalDate(), dateTime.toLocalTime());
+        System.out.printf("Started at: %s %s%n", dateTime.toLocalDate(), dateTime.toLocalTime());
+        Duration balancingDuration = Duration.between(startedAtInstant, OffsetDateTime.now());
+        System.out.printf("Balancing duration: %s%n%n", getPrettyDuration(balancingDuration));
```
Why not use the `toString` method of `Duration` instead of writing our own?

juncevich replied:
I think '1h 30m' better than 'PT1H30M'.

siddhantsangwan commented on hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto:
```diff
@@ -628,19 +628,20 @@ message ContainerBalancerStatusInfo {
 message ContainerBalancerTaskIterationStatusInfo {
```
Note that all other messages in this file have the suffix "Proto". This really improves code readability because the reader can distinguish between the java class `ContainerBalancerTaskIterationStatusInfo` and the java class that's generated by protoc for the proto message.

siddhantsangwan commented on hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/util/DurationUtilTest.java:
```diff
@@ -0,0 +1,93 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+class DurationUtilTest {
```
```suggestion
class TestDurationUtil {
```

siddhantsangwan reviewed (commented):
@juncevich thanks for all the updates. It looks like we still don't have unit tests on the server side for testing data that we're tracking in balancer, such as 
```
        iterationNumber,
        currentIterationResult.name(),
        iterationDuration
```
Please add tests for these, and also for any other important logic introduced in this pull request. I see that you've also been considering various edge cases - it'd be good to have tests for those cases to confirm that they've been resolved. For example, what happens if status info is called after a balancing iteration has been completed but the next one hasn't started yet? What if it's called multiple times during this time?

Better testing will give reviewers confidence that this change is correct.

siddhantsangwan commented:
@juncevich please let me know when you've addressed comments and this is ready for another review.

juncevich commented:
> please let me know when you've addressed comments and this is ready for another review.

@siddhantsangwan I've not understood about comments, can you explain about? PR is ready for another review.

juncevich commented:
@siddhantsangwan, can you take another look at this PR?

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerMetrics.java:
```diff
@@ -40,6 +40,9 @@ public final class ContainerBalancerMetrics {
       " in the latest iteration.")
   private MutableCounterLong dataSizeMovedGBInLatestIteration;
```
For backward compatibility.

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -306,83 +323,123 @@ private void balance() {
+    );
+    IterationInfo iterationInfo = new IterationInfo(
+                      0,
+        null,
```
Yep. Will fix this.

siddhantsangwan commented:
@juncevich if you've addressed all comments and fixed tests, I can start the CI workflow.

juncevich commented:
> @juncevich if you've addressed all comments and fixed tests, I can start the CI workflow.

Great, thanks! Just fixed the flaky test. Everything should be OK.

juncevich commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -306,83 +323,123 @@ private void balance() {
+    );
+  }
+
+  private ContainerBalancerTaskIterationStatusInfo getFilledCurrentIterationStatistic(int lastIterationNumber,
```
No, because `saveIterationStatistic` adds an item to iterationsStatistic. saveIterationStatistic contains information about finished statistic iterations. Current iteration - iteration in progress, not finished.

siddhantsangwan replied:
It's doable with a little refactoring. I hope we can reuse it so that we can keep things simple and increase maintainability.

We can start from something like this:
```
Subject: [PATCH] reuse
---
Index: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java b/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java
--- a/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java	(revision 6acad2b4fac6cf58e6d1c930e83bc86c14ccec10)
+++ b/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java	(date 1732537488628)
@@ -264,7 +264,7 @@
       }
 
       IterationResult iR = doIteration();
-      saveIterationStatistic(i + 1, iR);
+      iterationsStatistic.offer(getIterationStatistic(i + 1, iR.name(), getCurrentIterationDuration()));
 
       isCurrentIterationInProgress.compareAndSet(true, false);
 
@@ -314,15 +314,16 @@
     tryStopWithSaveConfiguration("Completed all iterations.");
   }
 
-  private void saveIterationStatistic(Integer iterationNumber, IterationResult currentIterationResult) {
-    long iterationDuration = now().toEpochSecond() - currentIterationStarted.toEpochSecond();
+  private ContainerBalancerTaskIterationStatusInfo getIterationStatistic(Integer iterationNumber,
+                                                                               String currentIterationResult,
+                                                                               long iterationDuration) {
     Map<UUID, Long> sizeEnteringDataToNodes =
         convertToNodeIdToTrafficMap(findTargetStrategy.getSizeEnteringNodes());
     Map<UUID, Long> sizeLeavingDataFromNodes =
         convertToNodeIdToTrafficMap(findSourceStrategy.getSizeLeavingNodes());
     IterationInfo iterationInfo = new IterationInfo(
         iterationNumber,
-        currentIterationResult.name(),
+        currentIterationResult,
         iterationDuration
     );
     ContainerMoveInfo containerMoveInfo = new ContainerMoveInfo(metrics);
@@ -332,12 +333,8 @@
         sizeEnteringDataToNodes,
         sizeLeavingDataFromNodes
     );
-    ContainerBalancerTaskIterationStatusInfo iterationStatistic = new ContainerBalancerTaskIterationStatusInfo(
-        iterationInfo,
-        containerMoveInfo,
-        dataMoveInfo
-    );
-    iterationsStatistic.offer(iterationStatistic);
+    //    iterationsStatistic.offer(iterationStatistic);
+    return new ContainerBalancerTaskIterationStatusInfo(iterationInfo, containerMoveInfo, dataMoveInfo);
   }
 
   private Map<UUID, Long> convertToNodeIdToTrafficMap(Map<DatanodeDetails, Long> nodeTrafficMap) {
@@ -375,7 +372,7 @@
     long iterationDuration = getCurrentIterationDuration();
 
     if (isCurrentIterationInProgress.get()) {
-      return getFilledCurrentIterationStatistic(lastIterationNumber, iterationDuration);
+      return getIterationStatistic(lastIterationNumber + 1, null, iterationDuration);
     } else {
       return getEmptyCurrentIterationStatistic(iterationDuration);
     }

```
One test is failing with this change - I have not checked why, but I think this patch can be a good starting point and help you.

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:
```diff
@@ -306,83 +323,123 @@ private void balance() {
-        .orElse(0);
+    if (isCurrentIterationInProgress.get()) {
+      return getFilledCurrentIterationStatistic(lastIterationNumber, iterationDuration);
+    } else {
+      return getEmptyCurrentIterationStatistic(iterationDuration);
+    }
```
It's critical in the sense that it's hard to maintain and understand. If someone wants to make changes to this area a few months from now, they'll have to figure out why they're getting three objects when balancer ran for only two iterations. I think we're better off removing this logic.

juncevich replied:
Ok, I will try to rewrite.

juncevich commented:
@siddhantsangwan I've fixed your review notices. Could you look through the changed code? I have to fix the robot test and PR will be ready.

juncevich commented:
@siddhantsangwan robot test fixed. PR is ready for another review round.

juncevich commented:
> @juncevich Thanks for updating. However, please try not to force push changes as that makes it hard to review the updates.

@siddhantsangwan I respect your point, but I can't entirely agree with it. You can choose the last commit with all fixes and without 'dirty' changes(commits). Force push makes the commit history more readable.
https://github.com/apache/ozone/pull/7139/commits/2e60bf3a340438607dd706ef0cad052853ae56be

adoroszlai commented on hadoop-ozone/dist/src/main/smoketest/balancer/testBalancer.robot:
```diff
@@ -170,6 +177,8 @@ Verify Container Balancer for RATIS/EC containers
 
     Run Balancer Verbose Status
 
+    Sleep                   40000ms
```
Please avoid fix amount of sleep.  Use `Wait Until Keyword Succeeds` or other retry logic.

juncevich replied:
Thanks. Will reopen PR and fix. It seems it is flaky.

juncevich replied:
Fixed in https://github.com/apache/ozone/pull/7569

adoroszlai reviewed (commented):
Can you please check test failure?

```
Verify Container Balancer for RATIS/EC containers                     | FAIL |
'ContainerBalancer is Not Running.' does not contain 'ContainerBalancer is Running.'
```

https://github.com/apache/ozone/actions/runs/12297371946/job/34319100286#step:7:14

ashishkumar50 commented:
`TestContainerBalancerStatusInfo.testGetCurrentStatisticsWhileBalancingInProgress` seems failing.
https://github.com/apache/ozone/actions/runs/12313619015/job/34368477497?pr=7551

juncevich commented:
> https://github.com/apache/ozone/actions/runs/12313619015/job/34368477497?pr=7551

Thanks. It is still flaky. Will check.

adoroszlai commented:
Created HDDS-11927 for the flaky test.

juncevich commented:
Added PR with fix HDDS-11927: https://github.com/apache/ozone/pull/7579
